### PR TITLE
Clean up CRD conversion code structure

### PR DESCRIPTION
**Pull request by ncdc**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Refactor the CRD conversion code:

* Rename crConverter to delegatingCRConverter
* Make creating a delegatingCRConverter a package function NewDelegatingConverter
* Move list/single item handling from the webhook converter into delegatingCRConverter
* Rename crConverterInterface to CRConverter so it is exported
* Change the CRConverter interface's input/output to be *unstructured.UnstructuredList
* Create and export NewNOPConverter

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ncdc** [comment]
@dims thx for the retest but I have some things I probably need to fix to get CI passing 😄

**Comment by ncdc** [comment]
Not sure if this e2e failure is from my changes ... let's try:

/retest

**Review by stevekuznetsov — APPROVED** [review]
Each commit here makes sense to me as a cleanup/clarification

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:689]: Is this used/useful? Would it be better to return an error if the provided `converterFactory` is nil?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:198]: nit: break out into utility function, e.g. `toUnstructuredList(in runtime.Object) (list *unstructured.UnstructuredList, isList bool)` ?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:352]: It is particularly noteworthy that apiVersion is not being compared here. Maybe focus on what the validation is achieving instead of which fields are excluded? (the full list of excluded fields is rather long)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:392]:
```go
var convertedMetaData map[string]interface{}
if ... {
  convertedMetaData = map[string]interface{}{}
} else {
   convertedMetaData = runtime.DeepCopyJSONValue(original.Object["metadata"])
}
converted.Object["metadata"] = convertedMetaData
```

To avoid the case and !ok error handling?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:246]: I was expecting `len(convertedObjects.Items)` to match `len(list)`, not `len(objectsToConvert)` (i.e. I was expecting any converter to pass-thru objects already in the correct version.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:240]: Is there an advantage to pre-checking this instead of just passing the data to the converter and letting it handle these cases?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:268]: Compare all of GroupVersionKind instead of doing one check for GroupVersion and another for Kind?

Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:268]: Also, put this logic into `validateConvertedObject`?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:427]: I was surprised to find validation code in this routine from the name/godoc.

Can the validation be split out? Maybe put this in `validateConvertedObject`? Something like `if !DeepEquals(in.GetLabels(), out.GetLabels()) { metav1validation.ValidateLabels(...) }` ?

I realize that there is a efficiency gain to combining the validation with the restore code, but I found it a more tedious to review and check for correctness this way..

**Comment by ncdc** [comment]
@jpbetz thanks for the review! Many/most of your comments are on code that was mostly a move from one file to another. Would you be ok if we considered handling comments on moved code separately (i.e. in a different PR, if needed)?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:240]: Code move from https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go#L251-L262

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:246]: Code move from https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go#L287

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:268]: Code move from https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go#L308-L313

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:275]: Code move from https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go#L319-L324

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:392]: Code move from https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go#L393-L405

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:427]: Code move from https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/webhook_converter.go#L440

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:240]: Oops, I missed that this whole section was just a move. Please feel free to close all my comments about it.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:246]: Oops, feel free to mark this as resolved.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:268]: Oops, feel free to mark this as resolved.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:275]: Oops, feel free to mark this as resolved.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:352]: Oops, feel free to mark this as resolved.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:392]: Oops, feel free to mark this as resolved.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:427]: Oops, feel free to mark this as resolved.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:689]: Hmm, thinking about this more, this method is part of serving a request, so you'd get the error when you interact with CRDs/CRs. I wonder if it would make more sense to check for this earlier, in `NewCustomResourceDefinitionHandler`?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (jpbetz) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:198]: I'm okay with it.

**Comment by jpbetz** [comment]
/lgtm

@deads2k or @liggitt would either you be willing to do an approver pass?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:64]: are callers going to get bitten by this returning a convertor that *has* to be wrapped by a separate call to NewDelegatingConverter to work properly?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:689]: +1 to returning an error in NewCustomResourceDefinitionHandler if converterFactory is nil, and assuming this is non-nil here, rather than turning wiring errors into maybe-data-destroying no-op conversions or request-time errors

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: this should happen after all the scale special cases and validversions checks, just before delegating to `c.converter.Convert`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:64]: Maaaaybe but the way this code is currently factored, this was the best "least" amount of work I could come up with.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go]: Fixed

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go
Comment (ncdc) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go]: Fixed

**Comment by liggitt** [comment]
/approve

lgtm once squashed

**Comment by ncdc** [comment]
@liggitt squashed

**Comment by ncdc** [comment]
Flake in `pkg/controller/volume/attachdetach/reconciler Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce`

/retest

**Pull request merged**
